### PR TITLE
move util/metrics to component-base

### PR DESCRIPTION
**Pull request by yastij**
Signed-off-by: Yassine TIJANI <ytijani@vmware.com>

**What type of PR is this?**

/kind cleanup
/priority important-soon

**What this PR does / why we need it**: moves `util/metrics` to the component-base repository

**Which issue(s) this PR fixes**: part of #81172

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by neolit123** [comment]
just noting i find the current state of component-base depending on prometheus a bit odd.
i was under the impression that the repository will serve a different purpose - hosting a state that all components will depend on.
metrics on the other hand doesn't feel like something that all components need and it pulls  prometheus.

with this PR it seems we are solidifying the binding to metrics / prometheus in that repository.

this PR and the series after it completely slipped of the radar cluster-lifecycle:
https://github.com/kubernetes/kubernetes/pull/77037

cc @logicalhan @sttts

**Comment by logicalhan** [comment]
> just noting i find the current state of component-base depending on prometheus a bit odd.
> i was under the impression that the repository will serve a different purpose - hosting a state that all components will depend on.
> metrics on the other hand doesn't feel like something that all components need and it pulls prometheus.
> 
> with this PR it seems we are solidifying the binding to metrics / prometheus in that repository.
> 
> this PR and the series after it completely slipped under the radar cluster-lifecycle:
> #77037
> 
> cc @logicalhan @sttts

Metrics is something all components need (except client-go). All the kubernetes binaries are currently instrumented with prometheus clients and as such, prometheus is already therefore a de-facto dependency of them. If we decide to move off of using prometheus for kubernetes, then we now have a single place where we can make that change. 

Also, moving the metrics stuff into component-base was actually a suggestion of @sttts.

**Comment by neolit123** [comment]
> Metrics is something all components need

certainly not deployers like kubeadm or in the future the "addon installer" which will replace the addon manager in /cluster for a wider use in the ecosystem.

AFAIK, the scope of component-base was just to host a set of common utils for CLI and component-config.

https://github.com/kubernetes/component-base#purpose

cc @mtaufen @stealthybox

**Comment by logicalhan** [comment]
> https://github.com/kubernetes/component-base#purpose

We should probably update that document.

**Comment by logicalhan** [comment]
> certainly not deployers like kubeadm or in the future the "addon installer" which will replace the addon manager in /cluster for a wider use in the ecosystem.

This is true. I'm not sure that that is sufficient reason to exclude something from component-base however.

> AFAIK, the scope of component-base was just to host a set of common utils for CLI and component-config.

According to the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/wgs/0032-create-a-k8s-io-component-repo.md#timeframe-and-implementation-order), under point 4 it states:

> Move server-related util packages from k8s.io/kubernetes/pkg/util/ to k8s.io/component-base/server. e.g. delegated authn/authz "/configz", "/healthz", and "/metrics" packages are suitable) 

An implication of this would be introducing a dependency on prometheus anyway.

**Comment by logicalhan** [comment]
/hold

(also gofmt, import-restrictions are failing)

I don't think we should move this until #81173 merges. After it merges, then the move should also migrate the `rate_limiter_use` metric to the stability framework (it will continue to work with non-migrated clients).

**Comment by neolit123** [comment]
> According to the KEP, under point 4 it states:

looks like i missed this part of the KEP.

**Comment by logicalhan** [comment]
/hold cancel

#81173 is merged. Would you mind moving this in its own directory under `https://github.com/kubernetes/component-base/tree/master/metrics/prometheus`? This will allow you to move and migrate these simultaneously.

**Comment by yastij** [comment]
@logicalhan - sgtm

**Comment by sttts** [comment]
> just noting i find the current state of component-base depending on prometheus a bit odd.
> i was under the impression that the repository will serve a different purpose - hosting a state that all components will depend on.
> metrics on the other hand doesn't feel like something that all components need and it pulls prometheus.

component-base is not about state.

And it is not about command line tools, it's about server components (daemons), always was. So metrics fit perfectly.

**Comment by alejandrox1** [comment]
@yastij any updates on this PR?

**Comment by yastij** [comment]
@alejandrox1 - I'll update this tomorrow

**Comment by alejandrox1** [comment]
forgot,
/remove-area e2e-test-framework
/remove-sig testing
the only part that is related to the e2e framework is a change in the import restrictions (which still needs work)

**Comment by liggitt** [comment]
/assign @logicalhan 
/unassign

(feel free to assign back to me once this has lgtm if approval is required)

**Review by logicalhan — APPROVED** [review]
/lgtm

**Comment by logicalhan** [comment]
needs rebase

**Comment by yastij** [comment]
rebased 

/assign @logicalhan @liggitt

File: cmd/kubeadm/.import-restrictions
Comment (liggitt) [inline cmd/kubeadm/.import-restrictions]: I don't think this makes sense in a list of imports that match `"SelectorRegexp": "k8s[.]io/kubernetes/pkg",`... can probably just drop this

File: pkg/controller/.import-restrictions
Comment (liggitt) [inline pkg/controller/.import-restrictions]: same comment about dropping this line from a `"SelectorRegexp": "k8s[.]io/kubernetes/pkg",` section

File: cmd/kube-controller-manager/app/core.go
Comment (liggitt) [inline cmd/kube-controller-manager/app/core.go]: verify is complaining this should be up one line

File: test/e2e/framework/.import-restrictions
Comment (liggitt) [inline test/e2e/framework/.import-restrictions]: drop this line from the `"SelectorRegexp": "k8s[.]io/kubernetes/pkg/",` section

File: pkg/util/metrics/OWNERS
Comment (liggitt) [inline pkg/util/metrics/OWNERS]: move this owners file with the package?

**Comment by liggitt** [comment]
approved, a few nits, will leave lgtm to @logicalhan

**Pull request merged**
